ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,243 @@
+        correlationIdToGetRequestInfo.entrySet().iterator();
+    while (inFlightRequestsIterator.hasNext()) {
+      Map.Entry<Integer, GetRequestInfo> entry = inFlightRequestsIterator.next();
+      if (time.milliseconds() - entry.getValue().startTimeMs > routerConfig.routerRequestTimeoutMs) {
+        operationTracker.onResponse(entry.getValue().replicaId, false);
+        setOperationException(
+            new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut));
+        inFlightRequestsIterator.remove();
+      } else {
+        // the entries are ordered by correlation id and time. Break on the first request that has not timed out.
+        break;
+      }
+    }
+
+    checkAndMaybeComplete();
+    if (isOperationComplete()) {
+      return;
+    }
+
+    Iterator<ReplicaId> replicaIterator = operationTracker.getReplicaIterator();
+    while (replicaIterator.hasNext()) {
+      ReplicaId replicaId = replicaIterator.next();
+      String hostname = replicaId.getDataNodeId().getHostname();
+      Port port = new Port(replicaId.getDataNodeId().getPort(), PortType.PLAINTEXT);
+      GetRequest getRequest = createGetRequest(blobId, getOperationFlag());
+      RequestInfo request = new RequestInfo(hostname, port, getRequest);
+      int correlationId = getRequest.getCorrelationId();
+      correlationIdToGetRequestInfo.put(correlationId, new GetRequestInfo(replicaId, time.milliseconds()));
+      requestRegistrationCallback.registerRequestToSend(this, request);
+      replicaIterator.remove();
+    }
+  }
+
+  /**
+   * Handle the body of the response: Deserialize and set the {@link BlobInfo} to return.
+   * @param payload the body of the response.
+   * @throws IOException if there is an IOException while deserializing the body.
+   * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
+   */
+  void handleBody(InputStream payload)
+      throws IOException, MessageFormatException {
+    if (blobInfo == null) {
+      blobInfo = new BlobInfo(MessageFormatRecord.deserializeBlobProperties(payload),
+          MessageFormatRecord.deserializeUserMetadata(payload).array());
+    } else {
+      // If the successTarget is 1, this case will never get executed.
+      // If it is more than 1, then, different responses will have to be reconciled in some way. Here is where that
+      // would be done. Since the store is immutable, currently we handle this by ignoring subsequent responses
+      // (if the successTarget is > 1).
+    }
+  }
+
+  /**
+   * Handle the given {@link ResponseInfo} and update the status of the operation.
+   * Based on the status of the response, this involves handling the body, notifying the operation tracker,
+   * notifying the response handler and possibly setting the operationException.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  @Override
+  void handleResponse(ResponseInfo responseInfo) {
```

Comment:
Yes. Moved out the actual handling of the `GetResponse` once it is extracted (and the other checks are completed) to a different method.